File: api/webhook.js
const express = require("express")
const bodyParser = require("body-parser")
const { handleMessage } = require("../src/messengerApi")
const { verifyWebhook } = require("../src/config")

const app = express()

app.use(bodyParser.json())

app.use((req, res, next) => {
  console.log(`${new Date().toISOString()} - ${req.method} ${req.url}`)
  console.log("Headers:", JSON.stringify(req.headers))
  console.log("Query:", JSON.stringify(req.query))
  console.log("Body:", JSON.stringify(req.body))
  next()
})

app.get("/api/webhook", (req, res) => {
  console.log("Requête GET reçue pour la vérification du webhook")
  return verifyWebhook(req, res)
})

app.post("/api/webhook", async (req, res) => {
  console.log("Début de la requête POST du webhook")
  console.log("Corps de la requête:", JSON.stringify(req.body))

  const body = req.body

  if (body.object === "page") {
    console.log("Événement de page reçu")
    for (const entry of body.entry) {
      console.log("Entrée reçue:", JSON.stringify(entry))
      if (entry.messaging) {
        const webhookEvent = entry.messaging[0]
        console.log("Événement Webhook reçu:", JSON.stringify(webhookEvent))

        const senderId = webhookEvent.sender.id
        console.log("ID de l'expéditeur:", senderId)

        if (webhookEvent.message) {
          console.log("Message reçu, appel de handleMessage")
          try {
            await handleMessage(senderId, webhookEvent.message)
            console.log("handleMessage terminé avec succès")
          } catch (error) {
            console.error("Erreur lors du traitement du message:", error)
          }
        } else if (webhookEvent.postback) {
          console.log("Postback reçu:", JSON.stringify(webhookEvent.postback))
          try {
            await handleMessage(senderId, { postback: webhookEvent.postback })
            console.log("handleMessage pour postback terminé avec succès")
          } catch (error) {
            console.error("Erreur lors du traitement du postback:", error)
          }
        } else {
          console.log("Événement non reconnu:", webhookEvent)
        }
      } else {
        console.log("Aucun événement de messagerie dans cette entrée")
      }
    }

    res.status(200).send("EVENT_RECEIVED")
  } else {
    console.log("Requête non reconnue reçue")
    res.sendStatus(404)
  }

  console.log("Fin de la requête POST du webhook")
})

app.use((err, req, res, next) => {
  console.error("Erreur non gérée:", err)
  res.status(500).send("Quelque chose s'est mal passé!")
})

module.exports = app

File: api/webhook.py
from flask import Flask, request, Response
import json
import os
from src.config import verify_webhook
from src.messenger_api import handle_message

app = Flask(__name__)

@app.route('/api/webhook', methods=['GET'])
def webhook_verification():
    """
    Endpoint pour la vérification du webhook par Facebook
    """
    print("Requête GET reçue pour la vérification du webhook")
    return verify_webhook(request)

@app.route('/api/webhook', methods=['POST'])
def webhook_handler():
    """
    Endpoint pour recevoir les événements du webhook
    """
    print("Requête POST reçue du webhook")
    data = request.json
    print(f"Corps de la requête: {json.dumps(data)}")
    
    if data.get('object') == 'page':
        print("Événement de page reçu")
        for entry in data.get('entry', []):
            print(f"Entrée reçue: {json.dumps(entry)}")
            messaging = entry.get('messaging', [])
            if messaging:
                webhook_event = messaging[0]
                print(f"Événement Webhook reçu: {json.dumps(webhook_event)}")
                
                sender_id = webhook_event.get('sender', {}).get('id')
                print(f"ID de l'expéditeur: {sender_id}")
                
                if webhook_event.get('message'):
                    print("Message reçu, appel de handle_message")
                    try:
                        handle_message(sender_id, webhook_event.get('message'))
                        print("handle_message terminé avec succès")
                    except Exception as e:
                        print(f"Erreur lors du traitement du message: {str(e)}")
                elif webhook_event.get('postback'):
                    print(f"Postback reçu: {json.dumps(webhook_event.get('postback'))}")
                    try:
                        handle_message(sender_id, {'postback': webhook_event.get('postback')})
                        print("handle_message pour postback terminé avec succès")
                    except Exception as e:
                        print(f"Erreur lors du traitement du postback: {str(e)}")
                else:
                    print(f"Événement non reconnu: {webhook_event}")
            else:
                print("Aucun événement de messagerie dans cette entrée")
        
        return Response("EVENT_RECEIVED", status=200)
    else:
        print("Requête non reconnue reçue")
        return Response(status=404)

@app.errorhandler(Exception)
def handle_error(e):
    print(f"Erreur non gérée: {str(e)}")
    return Response("Quelque chose s'est mal passé!", status=500)

if __name__ == '__main__':
    app.run(debug=True)

